### PR TITLE
Expose metrics through JMX

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/CompressingProbeRenderer.java [hunk 1]
```diff
@@ -182,4 +209,14 @@ private void moveNext() {
             }
         };
     }
+
+    private static class MorePublicByteArrayOutputStream extends ByteArrayOutputStream {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/CompressingProbeRenderer.java]: We used this pattern somewhere else I think? Maybe should be a Util class?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/CompressingProbeRenderer.java [hunk 2]
```diff
@@ -49,18 +58,41 @@
 
     private static final short BINARY_FORMAT_VERSION = 1;
 
+    private final ILogger logger;
+    private ObjLongConsumer<byte[]> consumer;
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/CompressingProbeRenderer.java]: can be made final

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsRenderPlugin.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.metrics;
+
+public interface MetricsRenderPlugin {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsRenderPlugin.java]: can we extend ProbeRenderer instead (with default implementations for renderNoValue and renderException) and also add some javadoc for the methods?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/CompressingProbeRenderer.java (hunk 1 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/CompressingProbeRenderer.java]: I did it several times, but it was always removed. This is the only instance now in jet code. And as a rule of thumb, single instance doesn't make it to `Util`. Btw, we should also make it non-synchronized, i reuse the instance and i wouldn't bet the JVM is able to elide locking...

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsRenderPlugin.java (hunk 3 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsRenderPlugin.java]: We handle the exceptions in a common way. `renderException` is supposed to send the exception text to the target (to MC or to JMX bean). In both our cases we just log the exception, if each plugin was a `ProbeRenderer` itself, the exception would be logged twice. 

`renderNoValue` is used in very special case: when the metric was deregistered while iterating the metrics. There's no point in having it.

Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsRenderPlugin.java]: can we rename MetricsRendererPlugin -> MetricsConsumer and the methods appropriately? Rendering implies showing something visually and plugin implies dynamic discovery/registration..

**Comment by cangencer** [comment]
Also missing Spring and XML config

**Pull request merged**